torch/_dynamo/compiled_autograd.py

```diff
@@ -203,6 +207,38 @@ def post_acc_grad_hook(self, input, hook_id):
             self.bind_tensors_to_proxies(input, proxies)
         return input
 
+    # Note: [Compiled autograd and cudagraphs]
+    # Eager autograd backward implements scalars as 0-dim tensors, see DivBackward0::other_.
+    # When compiled autograd traces those nodes, it lifts the scalar tensors, resulting in a graph
+    # with some cpu 0-dim tensor inputs. To prevent the entire graph from skipping cudagraph, we move the
+    # scalars tensors to cuda.
+    # To simplify this fx pass, we make a few assumptions specific to the compiled autograd graph:
+    #   1. cpu tensor inputs never affect the output device type of their users,
+    #      e.g. aten.div inherits type from numerator, not denominator
+    #   2. cpu tensor inputs are used in ops that accept cuda tensors too
+    def move_graph_nodes_to_cuda(self, graph) -> List[int]:
+        inputs_on_cpu = []
+        nodes = list(graph.nodes)
+        assert nodes[0].target == "inputs"
+        inputs = nodes[0]
+        inputs_users = list(inputs.users.keys())
+        # the ordering of the nodes should always [inputs, sizes, hooks, getitem, getitem1, ...]
+        # where getitemi accesses inputs[i]
+        first_getitem_idx = 3
+        assert nodes[first_getitem_idx] == inputs_users[0]
+        last_getitem_idx = first_getitem_idx + len(inputs_users) - 1
+        assert nodes[last_getitem_idx] == inputs_users[-1]
+        for i, node in enumerate(inputs_users):
+            is_cpu = node.meta["val"].device.type == "cpu"
+            is_scalar = len(node.meta["val"].size()) == 0
+            if is_cpu and is_scalar:
+                # move cpu scalars to cuda in the graph
+                node.meta["val"] = node.meta["val"].cuda()
```

Comment:
I think we need some additional checks here.  This code could run on a CPU-only graph and start moving scalars to cuda.  The machine might not even have cuda.

Perhaps check that at least one other input to this op is on cuda already.

Reply:
There is an existing pass to move constructors to cuda: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/post_grad.py#L916

We could add an option so that it moves input, cpu_scalars as well. Then we after applying, we could remove the conversions and take those as indices to convert. 

What point are we at the graph here ? I assume it's functionalized because backward. but is decomposed aten IR yet ?

Reply:
> What point are we at the graph here ?

This is equivalent to user code. we call torch.compile on this GraphModule

> We could add an option so that it moves input, cpu_scalars as well

I'm not convinced we want to move graph inputs to cuda automatically. User may have code after calling the compiled function that still uses the inputs. In this case, the compiled autograd graph is the user code, and we have no need for these inputs after, so it is safe to move them.

Reply:
Do we know that all of the uses of the cpu_scalar dims are safe to use cuda tensors for? for instance if it is used in a custom op

Reply:
Hmm, maybe i'm not following but do we know that all of the uses of the cpu_scalar dims are safe to use cuda tensors for? for instance if it is used in a custom op. the other pass checks to see if all uses of a cpu constructor are aten/prims, and that all other inputs to a use can be moved as well

Reply:
I added the aten/prims check on users of the cpu scalar we want to move. The cases are:

- python autograd.Function installed by AOTAutograd: CA traces into these, so all scalar users are visible, and scalar is moved iff ATen/prim.
- user-defined python autograd.Function: CA lifts these, so users are hidden. This will show up as a non-ATen/prim user to the scalar and the scalar will not be moved to CUDA.
- C++ autograd.Function: CA traces, scalar is moved iff ATen/prim.
- Backward hooks: Only gradients are inputs to backward hooks, so this case is affected by cpu scalar tensors that require grad. Since we have a check for CUDA inputs in the graph, we would also need a mixed device graph to trigger this case. And we don't seem to support this in fake tensor propagation, I added `test_scalar_grad_mixed_device` and xfail to track it.